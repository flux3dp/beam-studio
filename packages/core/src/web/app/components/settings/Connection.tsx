import * as React from 'react';

import alert from 'app/actions/alert-caller';
import alertConstants from 'app/constants/alert-constants';
import browser from 'implementations/browser';
import Controls from 'app/components/settings/Control';
import i18n from 'helpers/i18n';
import SelectControl from 'app/components/settings/SelectControl';
import storage from 'implementations/storage';
import { StorageKey } from 'interfaces/IStorage';

interface Props {
  originalIP: string;
  guessingPokeOptions: { value: any, label: string, selected: boolean, }[];
  autoConnectOptions: { value: any, label: string, selected: boolean, }[];
  updateConfigChange: (id: StorageKey, newVal: any) => void;
}

function Connection({
  originalIP,
  guessingPokeOptions,
  autoConnectOptions,
  updateConfigChange,
}: Props): JSX.Element {
  const lang = i18n.lang;
  const checkIPFormat = (e: React.FocusEvent): void => {
    const me = e.currentTarget as HTMLInputElement;
    const ips = me.value.split(/[,;] ?/);
    const ipv4Pattern = /^\d{1,3}[.]\d{1,3}[.]\d{1,3}[.]\d{1,3}$/;

    for (let i = 0; i < ips.length; i += 1) {
      const ip = ips[i];
      if (ip !== '' && typeof ip === 'string' && ipv4Pattern.test(ip) === false) {
        me.value = originalIP;
        alert.popUp({
          id: 'wrong-ip-error',
          type: alertConstants.SHOW_POPUP_ERROR,
          message: `${lang.settings.wrong_ip_format}\n${ip}`,
        });
        return;
      }
    }
    updateConfigChange('poke-ip-addr', me.value);
  };

  return (
    <>
      <div className="subtitle">
        {lang.settings.groups.connection}
        <span className="info-icon-medium">
          <img
            src="img/info.svg"
            onClick={() => browser.open(lang.settings.help_center_urls.connection)}
          />
        </span>
      </div>
      <Controls label={lang.settings.ip}>
        <input
          id="ip-input"
          type="text"
          autoComplete="false"
          defaultValue={storage.get('poke-ip-addr')}
          onBlur={checkIPFormat}
        />
      </Controls>
      <SelectControl
        id="set-guessing-poke"
        label={lang.settings.guess_poke}
        options={guessingPokeOptions}
        onChange={(e) => updateConfigChange('guessing_poke', e.target.value)}
      />
      <SelectControl
        id="set-auto-connect"
        label={lang.settings.auto_connect}
        options={autoConnectOptions}
        onChange={(e) => updateConfigChange('auto_connect', e.target.value)}
      />
    </>
  );
}

export default Connection;
