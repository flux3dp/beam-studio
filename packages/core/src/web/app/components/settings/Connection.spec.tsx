import React from 'react';
import { create } from 'zustand';
import { fireEvent, render } from '@testing-library/react';

jest.mock('@core/helpers/i18n', () => ({
  lang: {
    settings: {
      auto_connect: 'Automatically select the only machine',
      groups: {
        connection: 'Connection',
      },
      guess_poke: 'Search for machine IP address',
      help_center_urls: {
        connection: 'https://support.flux3dp.com/hc/en-us/sections/360000302135',
      },
      ip: 'Machine IP Address',
      wrong_ip_format: 'Wrong IP Formats',
    },
  },
}));

const popUp = jest.fn();

jest.mock('@core/app/actions/alert-caller', () => ({ popUp }));

const open = jest.fn();

jest.mock('@core/implementations/browser', () => ({ open }));

const mockGetConfig = jest.fn();
const mockSetConfig = jest.fn();

const useSettingStore = create(() => ({ getConfig: mockGetConfig, setConfig: mockSetConfig }));

jest.mock('@core/app/pages/Settings/useSettingStore', () => ({ useSettingStore }));
jest.mock('./components/SettingSelect');
jest.mock('./components/SettingFormItem');

import Connection from './Connection';

describe('should render correctly', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('invalid ip address is given', async () => {
    mockGetConfig.mockReturnValue('192.168.2.2');

    const { container } = render(
      <Connection
        options={
          [
            { label: 'On', value: true },
            { label: 'Off', value: false },
          ] as any
        }
      />,
    );

    expect(container).toMatchSnapshot();
    expect(mockGetConfig).toHaveBeenCalledTimes(4);
    expect(mockGetConfig).toHaveBeenNthCalledWith(1, 'poke-ip-addr');

    const input = container.querySelector('input');

    fireEvent.change(input, { target: { value: '192.168.3.3;192.168.4.4;192.168.1111.111' } });
    fireEvent.blur(input);

    expect(popUp).toHaveBeenCalledTimes(1);
    expect(popUp).toHaveBeenNthCalledWith(1, {
      id: 'wrong-ip-error',
      message: 'Wrong IP Formats\n192.168.1111.111',
      type: 'SHOW_POPUP_ERROR',
    });
    expect(mockSetConfig).not.toHaveBeenCalled();

    fireEvent.click(container.querySelector('img'));
    expect(open).toHaveBeenCalledTimes(1);
    expect(open).toHaveBeenNthCalledWith(1, 'https://support.flux3dp.com/hc/en-us/sections/360000302135');

    const controls = container.querySelectorAll('.select-control');

    fireEvent.change(controls[0], { target: { value: false } });
    expect(mockSetConfig).toHaveBeenCalledTimes(1);
    expect(mockSetConfig).toHaveBeenNthCalledWith(1, 'guessing_poke', false);

    fireEvent.change(controls[1], { target: { value: true } });
    expect(mockSetConfig).toHaveBeenCalledTimes(2);
    expect(mockSetConfig).toHaveBeenNthCalledWith(2, 'auto_connect', true);
  });

  test('valid ip address is given', () => {
    mockGetConfig.mockReturnValue('192.168.1.1');

    const { container } = render(
      <Connection
        options={
          [
            { label: 'On', value: true },
            { label: 'Off', value: false },
          ] as any
        }
      />,
    );

    const input = container.querySelector('input');

    fireEvent.change(input, { target: { value: '192.168.3.3;192.168.4.4' } });
    fireEvent.blur(input);
    expect(popUp).not.toHaveBeenCalled();
    expect(mockSetConfig).toHaveBeenCalledTimes(1);
    expect(mockSetConfig).toHaveBeenNthCalledWith(1, 'poke-ip-addr', '192.168.3.3;192.168.4.4');
  });
});
