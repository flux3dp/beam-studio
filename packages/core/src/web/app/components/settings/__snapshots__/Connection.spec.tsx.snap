// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly invalid ip address is given 1`] = `
<div>
  <div
    class="subtitle"
  >
    Connection
    <span
      class="info-icon-medium"
    >
      <img
        src="img/info.svg"
      />
    </span>
  </div>
  <mock-control
    label="Machine IP Address"
  >
    <input
      autocomplete="false"
      id="ip-input"
      type="text"
      value="192.168.2.2"
    />
  </mock-control>
  <div>
    mock-select-control id:
    set-guessing-poke
    label:
    Search for machine IP address
    options:
    [{"value":"TRUE","label":"On","selected":true},{"value":"FALSE","label":"Off","selected":false}]
    <input
      class="select-control"
    />
  </div>
  <div>
    mock-select-control id:
    set-auto-connect
    label:
    Automatically select the only machine
    options:
    [{"value":"TRUE","label":"On","selected":false},{"value":"FALSE","label":"Off","selected":true}]
    <input
      class="select-control"
    />
  </div>
</div>
`;
